src/proof.rs

```diff
@@ -0,0 +1,337 @@
+// Copyright 2020 Kodebox, Inc.
+// This file is part of CodeChain.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use crate::nibbleslice::NibbleSlice;
+use crate::node::Node;
+use ccrypto::{blake256, BLAKE_NULL_RLP};
+use primitives::Bytes;
+use primitives::H256;
+
+// Unit of a proof.
+//#[derive(Clone, Eq, PartialEq, Debug, RlpEncodable, RlpDecodable)]
+#[derive(Clone, Eq, PartialEq, Debug)]
+pub struct CryptoProofUnit<H, K, V> {
```

Comment:
I still have a question on whether we need to make this type generic.
If our usage doesn't need it to be generic, this merely makes the user hard to use it.
Consider it again that we are not implementing a general cryptographic library, but a library for use in the ICS of Foundry.
Do we need to make them generic parameters?

Reply:
Ok now I understand. Then is it ok to move all high-level abstraction to Foundry (which has high chance of implementing proof of other cryptographic data structure, since there might be some supports on other chain for the ICS), and keep only verification/proof creation logic here? If then all Foundry has to do is just implementing those traits by just trivially calling the logic-only module left here. It will make this repo very specific to certain structure. (the TrieDB)

Reply:
@majecty and I had short discussion, and concluded that it will be good.

Reply:
Why should we move a high-level abstraction to Foundry?
Isn't it possible to implement the proof for the concrete type here and make Foundry call it?

Reply:
Then you want to only remove the generics but still keep the high-level abstraction?

Reply:
What is the abstraction you meant?
I think there isn't a need to abstract it.
Why not implement the proof and verification for the specific type and make Foundry use it?

Reply:
Can I have a hangout session with you?

Reply:
We (me, @majecty , @sgkim126 ) concluded that it's better to keep only concrete routine here.